### ISSUE TITLE
why can not pickle lambda in spark but useful in pure python code.

### ISSUE THREAD

**Issue by 143230** [cython/cython#2584]
i submit my spark job using pure python like below:
sc.textFile(dp)..map(lambda line: line.split(' '))
it is okay and perfect,
however, when using cython and change it into .so file, the spark raise exceptions and told
me that can not pickle lambda function.

help me, thanks

**Comment by robertwb** [comment]
Spark pickles lambdas by capturing the bytecode/closure and re-creating it on the other end. There's no bytecode and the closure may be non-Python objects for Cython-defined lambdas. However, it is a valid feature request to make this possible by implementing their `__reduce__` methods.

**Comment by scoder** [comment]
Some more details: Python's pickle module needs to be able to find the code of the function somehow. In the case of a Python function, that can be done by pickling the code object that contains the Python byte code. In the case of a compiled (binary/native) function, there is no byte code, and pickling native code isn't very useful/portable/easy. Thus, for native functions, the pickle module would want to pickle the state of the function and the import path to find the function. Locally defined (nested) functions cannot be imported, so pickle will not be able to unpickle their state.

That being said, in some cases, it might be possible to register a global function in the module as entry point for unpickling Cython compiled functions. But that's a) probably quite some work and b) only applicable to functions for which the entire state can be pickled, i.e. no pointer variables or C++ objects locally or in any of the outer closures. Platform specific data types also play a role here. It would be a challenge to pickle a C long value on a platform with 64bits and unpickling it on a platform with 32bit long integers, for example.

**Comment by robertwb** [comment]
I think the way to go about this is to write a custom __reduce__ and leverage pickling the closure class just as we do for other cdef classes.

**Comment by will-ca** [comment]
I'm not sure whether this would be the same issue or require a separate implementation, but getting Cythonized objects to work well with `dill.dumps()` would be quite nice as well.

**Comment by jakirkham** [comment]
Did anyone try with cloudpickle?

**Comment by ash-ishh** [comment]
@jakirkham cloudpickle is not working

**Comment by nagesh-chowdaiah** [comment]
@scoder  any udpates on the workout to fix the issue from cython side?  we need this feature very badly

**Comment by da-woods** [comment]
@nagesh-chowdaiah (since you tagged me in the other issue). I have no immediate plans to work on this.

If you or anyone else want to have a go at this: CyFunctions store their closure as a cdef class, therefore you can use the code that already exists in AnalyseDeclarationsTransform to pickle that cdef class (I don't think that's 100% simple because the closure cdef class is generated quite late). You'd then just need to write fairly simple `__reduce__` function for `CyFunction`.

If you don't want to do that then I'd suggest you don't use a nested function: instead write a pickleable (`cdef`?) `class` with a `__call__` method and use that as your callable:

```
def old_version(a):
    def old_version_inner():
       return a
```

gets changed to

```
class NewVersion:
   def __init__(self, a):
      self.a = a
   def __call__(self):
      return self.a
   # possible extra functions to make it pickleable...
```

**Comment by da-woods** [comment]
I'm having a tentative look at how to do this. (This should not be taken as a guarantee it'll be implemented soon!)

I think there's at least one small design choice: do we use the `auto_pickle` and `auto_pickle_force` directives to control this or invent new ones? `auto_pickle` seems pretty safe. `auto_pickle_force` risks breaking existing code that uses it but only intends to apply it to classes. On the other hand, using the existing directives makes sense to me.

**Comment by scoder** [comment]
IMHO, pickling lambdas and inner functions is a special use case that most people don't have. And even those who do, usually only want to pickle very few of them and not all. This does not seem worth bloating the module namespace with tons of closure classes to make them available to `pickle`, because even `pickle` won't need to see most of them. But I think we can get away with a single entry point, as follows.

- Each module that supports pickling inner functions provides a Python visible function as single entry point. It will be visible at the module level, so that `pickle` can use it. Let's name it `__pyx_cyfunction_unpickle`.
- `CyFunction` implements `__pickle__()` to return a 2-tuple with that function as first item and then an argument tuple as second item, with two elements:
  - a reference (a name string? an ID number?) for the closure class type of the specific inner function
  - the state struct of the closure class, as picklable object (as provided by the closure's `__reduce_cython__` method)
- On unpickling, `pickle` will call the module's unpickle function with those two arguments, which would then look up the right closure class and instantiate it using its `__setstate_cython__` method.

For the closure reference, a name string (preferably the `qualname`) is slower but seems more helpful than a (simple increasing) ID number since it would allow more (compatible) changes to the module without invalidating the pickles. I don't think there is a good way to make lambdas compatible across module versions, since they inherently aim to have no identifiable name. One might think that hashing their code could work, but then what if two lambdas inside the same outer function context have the same code? I think relying on the `qualname` seems good enough for them, too.

With this design, I think we can afford to include inner functions and lambdas in `auto_pickle`. It bloats the module size, but it won't get in the way otherwise. And users can mitigate the size effect by using `@auto_pickle(True/False)` as a targeted decorator, also for inner (and outer) functions and specific extension types.

**Comment by da-woods** [comment]
Large chunks of that are pretty close to the design I was heading towards. Probably the main difference is that you're proposing to only generate `__reduce_cython__` and `__set_state_cython__` for the closure classes and not actually make them fully pickleable. That seems reasonable to me.

> CyFunction implements __pickle__() to return a 2-tuple with that function as first item and then an argument tuple as second item, with two elements:
>  *  a reference (a name string? an ID number?) for the closure class type of the specific inner function
>  *  the state struct of the closure class, as picklable object (as provided by the closure's __reduce_cython__ method)

I was going to use the function cname as the reference (on the basis that it's guaranteed to be unique per module otherwise C breaks and it's convenient for code generation). I think the state might need a few more elements: possibly the `classobj` if it exists, and maybe `self` for fused functions.

>  I don't think there is a good way to make lambdas compatible across module versions, since they inherently aim to have no identifiable name. [...]

And the cname is presumably just as fragile. I imagine support for lambdas will have to be fairly tied to an individual module version.

**Comment by da-woods** [comment]
> > * a reference (a name string? an ID number?) for the closure class type of the specific inner function
> > [...]
> 
> I was going to use the function cname as the reference (on the basis that it's guaranteed to be unique per module otherwise C breaks and it's convenient for code generation).

Just a couple more comments on this: I think the function cname requires an extra module-level support function. On the other hand, the qualname isn't unique

```
def f(select, b):
   if select:
     def g():
       return b
   else:
     def g():
       return -b
   return g
```

gives two functions with the same qualname

**Comment by robertwb** [comment]
Yes, we'd need some way to identify the two `g`s here. An incrementing suffix might be problematic as (for example) adding an `elif` clause in the example above would cause an off-by-one error. (FWIW, I've run into this request before as well, and it's unfortunate to tell people the only workaround is creating an explicit class or doing it in a Python file. But this has always been in the context of distributed processing (v.s. long term storage) so totally opaque identifiers would be sufficient.)
